src/rtprecv.c

```diff
@@ -0,0 +1,507 @@
+	struct rtp_sock *rtp;          /**< RTP Socket                       */
+	stream_pt_h *pth;              /**< Stream payload type handler      */
+	stream_rtp_h *rtph;            /**< Stream RTP handler               */
+	stream_rtpestab_h *rtpestabh;  /**< RTP established handler          */
+	void *arg;                     /**< Stream argument                  */
+	void *sessarg;                 /**< Session argument                 */
+	int pt;                        /**< Previous payload type            */
+};
+
+
+static int lostcalc(struct rtp_receiver *rx, uint16_t seq)
+{
+	const uint16_t delta = seq - rx->pseq;
+	int lostc;
+
+	if (rx->pseq == (uint32_t)-1)
+		lostc = 0;
+	else if (delta == 0)
+		return -1;
+	else if (delta < 3000)
+		lostc = delta - 1;
+	else if (delta < 0xff9c)
+		lostc = 0;
+	else
+		return -2;
+
+	rx->pseq = seq;
+	return lostc;
+}
+
+
+static int handle_rtp(struct rtp_receiver *rx, const struct rtp_header *hdr,
+		      struct mbuf *mb, unsigned lostc, bool drop)
+{
+	struct rtpext extv[8];
+	size_t extc = 0;
+	bool ignore = drop;
+
+	/* RFC 5285 -- A General Mechanism for RTP Header Extensions */
+	if (hdr->ext && hdr->x.len && mb) {
+
+		const size_t pos = mb->pos;
+		const size_t end = mb->end;
+		const size_t ext_stop = mb->pos;
+		size_t i;
+		int err;
+
+		if (hdr->x.type != RTPEXT_TYPE_MAGIC) {
+			debug("stream: unknown ext type ignored (0x%04x)\n",
+			     hdr->x.type);
+			goto handler;
+		}
+
+		size_t ext_len = hdr->x.len*sizeof(uint32_t);
+		if (mb->pos < ext_len) {
+			warning("stream: corrupt rtp packet,"
+				" not enough space for rtpext of %zu bytes\n",
+				ext_len);
+			return 0;
+		}
+
+		mb->pos = mb->pos - ext_len;
+		mb->end = ext_stop;
+
+		for (i=0; i<RE_ARRAY_SIZE(extv) && mbuf_get_left(mb); i++) {
+
+			err = rtpext_decode(&extv[i], mb);
+			if (err) {
+				warning("stream: rtpext_decode failed (%m)\n",
+					err);
+				return 0;
+			}
+		}
+
+		extc = i;
+
+		mb->pos = pos;
+		mb->end = end;
+	}
+
+ handler:
+	stream_stop_natpinhole(rx->strm);
+
+	rx->rtph(hdr, extv, extc, mb, lostc, &ignore, rx->arg);
+	if (ignore)
+		return EAGAIN;
+
+	return 0;
+}
+
+
+/**
+ * Decodes one RTP packet
+ *
+ * @param s The stream
+ *
+ * @return 0 if success, EAGAIN if it should be called again in order to avoid
+ * a jitter buffer overflow, otherwise errorcode
+ */
+static int decode_frame(struct rtp_receiver *rx)
+{
+	struct rtp_header hdr;
+	void *mb;
+	int lostc;
+	int err;
+	int err2;
+
+	if (!rx)
+		return EINVAL;
+
+	if (!rx->jbuf)
+		return ENOENT;
+
+	err = jbuf_get(rx->jbuf, &hdr, &mb);
+	if (err && err != EAGAIN)
+		return ENOENT;
+
+	lostc = lostcalc(rx, hdr.seq);
+
+	err2 = handle_rtp(rx, &hdr, mb, lostc > 0 ? lostc : 0, err == EAGAIN);
+	mem_deref(mb);
+
+	if (err2 == EAGAIN)
+		return err2;
+
+	return err;
+}
+
+
+void rtprecv_decode(const struct sa *src, const struct rtp_header *hdr,
+		     struct mbuf *mb, void *arg)
```

Comment:
if this is a callback handler, we should try to make them static and not expose it in api.